### DISCUSSION TITLE
HowTo: Create type safe cross pydantic version compatible config mixins

### DISCUSSION THREAD

**Discussion by ddluke** [pydantic/pydantic#9309] [Question]
Hi folks! :) 

## Description

We have started to add support for pydantic v2 in our distributed MLops platform. A core challenge we are currently facing, is that pydantic is literally part of every single python package in our platform and used for various use cases:

- microservices running fastapi + pydantic
- configuration management for training + processing jobs
- modeling structured log payloads
- and more

Our core issue is, that we need to support both versions for an unknown period. That is mainly, because there are ML models that have been trained with pydantic v1 and need to be served with v1, while there might be new ML models trained with v2 that also need to be served with v2. At runtime, especially in our ML endpoints, we might have either v1 or v2 installed, and we need to be able to support both.

This brings us into a position where some components of our platform will need to be compatible with both versions for an unknown time (could be weeks, months, ...). This creates the need to provide a cross version compatibility layer that makes transition easier. 

While we have been able to get most utilised pydantic features test and type safe cross version (using some sort of wrapper/adapter code), pydantic model configuration is a thing which still bothers us. This is because there might be hundreds/thousands of places where base models are persisted in code with different configuration settings (mutable or immutable, with aliasing or without, with extras or without, etc.).

To ease this transition, we have thought about creating cross version compatible mixin classes that can be requested to create configured pydantic models. This is somewhat the idea:

```python
# type checking this code does work via something like `python -m mypy foo.py --always-true PYDANTIC_V1`
import packaging.version
import pydantic

PYDANTIC_V1 = packaging.version.parse(pydantic.__version__).major == 1

if PYDANTIC_V1:
    class Immutable(pydantic.BaseModel):
        """mixin to make a pydantic model immutable"""

        class Config:
            allow_mutation = False
else:
    from pydantic import ConfigDict


    class Immutable(pydantic.BaseModel):
        """mixin to make a pydantic model immutable"""

        model_config = ConfigDict(frozen=True)

```

To make a model immutable, our customers could then just simply declare their models this way:
```python
from foo import Immutable

class JobConfiguration(Immutable):
    foo: str
```

However, this becomes tricky with multiple inheritance. We have tried two possible approaches with v1, which both fail type checks with the pydantic mypy plugin

## Approach No 1 (using Config classes)

```python
from typing import Any

from pydantic import BaseModel, Field


class Immutable(BaseModel):
    """mixin to make a pydantic model immutable"""

    class Config:
        allow_mutation = False


class WithAliases(BaseModel):
    """mixin to make a pydantic model to accept and serialize by alias"""

    class Config:
        allow_population_by_field_name = True

    def json(self, by_alias: bool = True, **kwargs: Any) -> str:
        return super().json(by_alias=by_alias, **kwargs)


class Customer(Immutable, WithAliases):
    id_: str = Field(..., alias="id")


instance = Customer(id_="foo")
instance.id_ = "1"
```

Running mypy on this code fails

```shell
python -m mypy v1.py --always-true PYDANTIC_V1 
v1.py:23: error: Definition of "Config" in base class "Immutable" is incompatible with definition in base class "WithAliases"  [misc]
v1.py:28: error: Property "id_" defined in "Customer" is read-only  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

## Approach No 2 (using class kwargs)

```python
from typing import Any

from pydantic import BaseModel, Field


class Immutable(BaseModel, allow_mutation=False):
    """mixin to make a pydantic model immutable"""


class WithAliases(BaseModel, allow_population_by_field_name=True):
    """mixin to make a pydantic model to accept and serialize by alias"""

    def json(self, by_alias: bool = True, **kwargs: Any) -> str:
        return super().json(by_alias=by_alias, **kwargs)


class Customer(Immutable, WithAliases):
    id_: str = Field(..., alias="id")


instance = Customer(id_="foo")
instance.id_ = "1"

```

Running mypy on this code fails as well, albeit with different error messages

```shell
python -m mypy v2.py --always-true PYDANTIC_V1
v2.py:21: error: Unexpected keyword argument "id_" for "Customer"; did you mean "id"?  [call-arg]
Found 1 error in 1 file (checked 1 source file)
```

Does anyone know of a way to use multiple configuration mixin classes in pydantic v1 and have them pass mypy type checks?

## System Details

```
Virtualenv
Python:         3.10.14
Implementation: CPython

System
Platform:   darwin
OS:         posix

Dependencies
mypy              1.9.0
mypy-extensions   1.0.0
packaging         24.0
pydantic          1.10.10
tomli             2.0.1
typing-extensions 4.11.0
```

## mypy configuration
```toml
[tool.mypy]
plugins = [
  "pydantic.mypy"
]
strict = true
implicit_reexport = true
ignore_missing_imports = true
check_untyped_defs = true
show_error_codes = true
warn_redundant_casts = true
warn_unused_configs = true
warn_unused_ignores = true
# for strict mypy: (this is the tricky one :-))
disallow_untyped_defs = true

[tool.pydantic-mypy]
init_forbid_extra = true
init_typed = true
warn_required_dynamic_aliases = true
warn_untyped_fields = true
```

**Comment by ybressler** [comment]
My team is doing something similar in our pydantic implementation. We've found success with 2 general approaches.

## Approach 1: Make models compatible with V1 and V2
This is more similar to what you're suggesting. This begins to break when doing fancy stuff, like custom types.

Implementation starts having many logic branches >> we found that the benefit wasn't worth the return on this, sadly. Instead, we went with approach 2.

## Approach 2: Downshift to V1.
Sanity purposes, it is easier to downshift to V1, and reference all pydantic entities from an upfront declaration.

For example, in `base.py`:
```py3
IS_PYDANTIC_VERSION_1: bool = ...  # some logic

if IS_PYDANTIC_VERSION_1:
    from pydantic import BaseModel
else:
    from pydantic.v1 import BaseModel
```
